raster-foundry/raster-foundry
# Decrease duplicated queries for tile rendering for thumbnails and mosaic TMS

## Overview

This PR reduces the number of scene by id queries in the thumbnail rendering process from four to one. It also _just about_ removes the need to have `ProjectStoreImplicits` for `SceneDao` at all, but they're still used in `MosaicImplicits`, so they can't _quite_ be removed yet.

### Checklist


## Notes

~While I'm working on this and #5119 simultaneously, it seemed worth it to break this out, since the change to the return type of `authorized` had far-reaching consequences~ now they're combined

It will probably be easiest to focus on the [last four commits](https://github.com/raster-foundry/raster-foundry/pull/5139/files/be68aed8b22aba0770eeafefb587a16397e3b19d..464fcc4bd94b1712215ec228a150b2f936590338) for the first change, then the [next two commits](https://github.com/raster-foundry/raster-foundry/pull/5139/files/9d65c53828e77cbb5300932166293de15404b349..310530fbfbdd58cf9d55c012745e70a3b9ebb661) for the second change

Here are timings from `develop` in its current state / without this PR:

https://s3.amazonaws.com/rasterfoundry-staging-it-us-east-1/gatling/341/index.html

The staging cluster scaled to its maximum number of task instances after the load test on `develop`, from starting at the minimum.

And here are timings from `test/js/dedupe-queries`, which is in the same state as this branch:

https://s3.amazonaws.com/rasterfoundry-staging-it-us-east-1/gatling/342/index.html

The staging cluster also scaled to its maximum number of instances after this load test was over, from starting at its minimum.

The two noticeable differences for me were:

- in the response time distribution, the right tail is considerably shorter on this branch than it is on develop
- in the response time percentiles over time, the mid-percentiles (50, 75, 80, 85, 90) stay far lower on this branch than on develop

To verify this, I scrolled to the same page position (thanks, standard reporting layout) and tabbed back and forth in the browser. It's not super scientific, but the differences are stark enough that they're noticeable even without a more careful reading.

## Testing Instructions

- enable database logging in the way you prefer (probably use the postgres container command option, since #5129 is still wip -- add `command: postgres -c log_statement=all` to your `postgres` service in docker-compose.yml)
- reassemble tile and api servers
- bring up your servers and frontend
- get a bearer token from the frontend
- request a thumbnail for the `Local File Single Scene` project:

```bash
$ http 'http://localhost:8081/scenes/533fdfa0-0f76-49d8-8114-e9326d6c91ad/thumbnail?floor=25&height=128&token=<YOUR TOKEN>'`
```

- confirm in your postgres logs that you only have one select from the `scenes` table
- request a tms tile for the planet project:

```bash
$ http 'http://localhost:8081/94bb99c1-c8da-4321-a03b-fce884009a90/15/6600/12846/?tag=1567026023147&token=<YOUR TOKEN>'
```

- confirm in your postgres logs that you have no more than two database queries for getting a project layer by id (you can end up with two because of parallelism / the cache not being populated at the time the second call occurs)
- confirm that you only have a single read from `scene_to_layers`

Closes #5119 

Closes #5122
